nav2_util/src/lifecycle_node.cpp

```diff
@@ -64,6 +73,24 @@ void LifecycleNode::createBond()
   bond_->start();
 }
 
+void LifecycleNode::on_rcl_preshutdown()
+{
+  RCLCPP_INFO(
+    get_logger(), "Running Nav2 LifecycleNode rcl preshutdown (%s)",
+    this->get_name());
+  destroyBond();
+}
+
+void LifecycleNode::register_rcl_preshutdown_callback()
+{
+  rclcpp::Context::SharedPtr context = get_node_base_interface()->get_context();
```

Comment:
If something like this is truly necessary, I think we need to escalate this to rclcpp that we cannot get clean exits from component containers - this seems like a serious issue. 

Do you see this also with non-composed bringups (e.g. launching a server itself in its own process)? We have a `nav2_util` util that can help with transitioning up the nodes: https://github.com/ros-planning/navigation2/blob/main/nav2_util/src/lifecycle_bringup_commandline.cpp

Note also that https://github.com/ros2/rclcpp/pull/2085 was just merged, which should help with specifically the isolated component container shutdown. Though curious if you see this issue with any other component container type.

Reply:
Also, I love the debugging and learning about new APIs I hadn't yet even played with myself in ROS 2! The list of the lower-level APIs are so massive, I've just never bothered to even look at many of them. 

Edit: As I looked for it, or here's an example the autogenerated docs don't even have :sweat: